ios/browser/api/sync/brave_sync_worker.cc

```diff
@@ -0,0 +1,390 @@
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#import "brave/ios/browser/api/sync/brave_sync_worker.h"
+
+#include <string>
+#include <vector>
+
+#include "base/json/json_writer.h"
+#include "base/memory/weak_ptr.h"
+#include "base/scoped_observer.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/sys_string_conversions.h"
+#include "brave/components/brave_sync/brave_sync_prefs.h"
+#include "brave/components/brave_sync/crypto/crypto.h"
+#include "brave/components/sync/driver/brave_sync_profile_sync_service.h"
+#include "components/sync/driver/profile_sync_service.h"
+#include "components/sync/driver/sync_service.h"
+#include "components/sync/driver/sync_service_observer.h"
+#include "components/sync_device_info/device_info.h"
+#include "components/sync_device_info/device_info_sync_service.h"
+#include "components/sync_device_info/device_info_tracker.h"
+#include "components/sync_device_info/local_device_info_provider.h"
+#include "ios/chrome/browser/application_context.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/browser_state/chrome_browser_state_manager.h"
+#include "ios/chrome/browser/sync/device_info_sync_service_factory.h"
+#include "ios/chrome/browser/sync/profile_sync_service_factory.h"
+#include "ios/chrome/browser/sync/sync_setup_service.h"
+#include "ios/chrome/browser/sync/sync_setup_service_factory.h"
+#include "ios/web/public/thread/web_thread.h"
+
+namespace {
+static const size_t SEED_BYTES_COUNT = 32u;
+}  // namespace
+
+namespace {
+
+// A structure which contains all the configuration information for sync.
+struct SyncConfigInfo {
+  SyncConfigInfo();
+  ~SyncConfigInfo();
+
+  bool encrypt_all;
+  bool set_new_passphrase;
+};
+
+SyncConfigInfo::SyncConfigInfo()
+    : encrypt_all(false), set_new_passphrase(false) {}
+
+SyncConfigInfo::~SyncConfigInfo() {}
+
+// Return false if we are not interested configure encryption
+bool FillSyncConfigInfo(syncer::SyncService* service,
+                        SyncConfigInfo* configuration) {
+  bool first_setup_in_progress =
+      service && !service->GetUserSettings()->IsFirstSetupComplete();
+
+  configuration->encrypt_all =
+      service->GetUserSettings()->IsEncryptEverythingEnabled();
+
+  bool sync_prefs_passphrase_required =
+      service->GetUserSettings()->IsPassphraseRequired();
+
+  if (!first_setup_in_progress) {
+    if (!configuration->encrypt_all) {
+      configuration->encrypt_all = true;
+      configuration->set_new_passphrase = true;
+    } else if (sync_prefs_passphrase_required) {
+      configuration->set_new_passphrase = false;
+    } else {
+      return false;
+    }
+  }
+  return true;
+}
+
+}  // namespace
+
+BraveSyncDeviceTracker::BraveSyncDeviceTracker(std::function<void()> onDeviceInfoChanged) : onDeviceInfoChanged_(onDeviceInfoChanged) {
+  
+  ios::ChromeBrowserStateManager* browserStateManager =
```

Comment:
as mentioned in other comments, this is an anti-pattern. The DeviceInfoTracker should be passed into the constructor